internal/core/operations/newschedulerversion/new_scheduler_version_executor.go

```diff
@@ -0,0 +1,143 @@
+		schedulerManager: schedulerManager,
+	}
+}
+
+func (ex *CreateNewSchedulerVersionExecutor) Execute(ctx context.Context, op *operation.Operation, definition operations.Definition) error {
+
+	logger := zap.L().With(
+		zap.String("scheduler_name", op.SchedulerName),
+		zap.String("operation_definition", op.DefinitionName),
+		zap.String("operation_id", op.ID),
+	)
+	opDef, ok := definition.(*CreateNewSchedulerVersionDefinition)
+	if !ok {
+		return errors.NewErrInvalidArgument(fmt.Sprintf("invalid operation definition for %s operation", ex.Name()))
+	}
+
+	newScheduler := opDef.NewScheduler
+	currentActiveScheduler, err := ex.schedulerManager.GetActiveScheduler(ctx, opDef.NewScheduler.Name)
+	if err != nil {
+		logger.Error("error getting active scheduler", zap.Error(err))
+		return fmt.Errorf("error getting active scheduler: %w", err)
+	}
+
+	isSchedulerMajorVersion := ex.schedulerManager.IsMajorVersionUpdate(currentActiveScheduler, newScheduler)
+
+	err = ex.populateSchedulerNewVersion(newScheduler, currentActiveScheduler.Spec.Version, isSchedulerMajorVersion)
+	if err != nil {
+		return err
+	}
+
+	if isSchedulerMajorVersion {
+		err = ex.roomManager.ValidateGameRoomCreation(ctx, newScheduler)
+		if err != nil {
+			logger.Error("could not validate new game room creation", zap.Error(err))
+			return err
+		}
+	}
+
+	err = ex.createNewSchedulerVersion(ctx, newScheduler, logger)
+	if err != nil {
+		return err
+	}
+
+	switchActiveVersionOp, err := ex.enqueueSwitchActiveVersionOperation(ctx, newScheduler, logger)
+	if err != nil {
+		return err
+	}
+
+	logger.Info(fmt.Sprintf("%s operation succeded, %s operation enqueued to continue scheduler update process", opDef.Name(), switchActiveVersionOp.DefinitionName))
+	return nil
+}
+
+func (ex *CreateNewSchedulerVersionExecutor) OnError(ctx context.Context, op *operation.Operation, definition operations.Definition, executeErr error) error {
+	return nil
+}
+
+func (ex *CreateNewSchedulerVersionExecutor) Name() string {
+	return OperationName
+}
+
+func (ex *CreateNewSchedulerVersionExecutor) createNewSchedulerVersion(ctx context.Context, newScheduler *entities.Scheduler, logger *zap.Logger) error {
+	err := ex.schedulerManager.CreateNewSchedulerVersion(ctx, newScheduler)
+	if err != nil {
+		logger.Error("error creating new scheduler version in db", zap.Error(err))
+		return fmt.Errorf("error creating new scheduler version in db: %w", err)
+	}
+	return nil
+}
+
+func (ex *CreateNewSchedulerVersionExecutor) enqueueSwitchActiveVersionOperation(ctx context.Context, newScheduler *entities.Scheduler, logger *zap.Logger) (*operation.Operation, error) {
+	switchActiveVersionOp, err := ex.schedulerManager.EnqueueSwitchActiveVersionOperation(ctx, newScheduler)
+	if err != nil {
+		// TODO(guilhermbrsp): Maybe we should rollback the creation of the new scheduler version if some error happens here
+		logger.Error("error enqueuing switch active version operation", zap.Error(err))
```

Comment:
@lyracampos @caiordjesus What do you guys think about this? should we rollback the creation of new scheduler here or in the OnError method?

Reply:
Just for record: As we discussed before, we should open a pg transaction before creating the new version, create it using the transaction, and only commit after everything works well (and rollback if anything goes wrong)

Reply:
@GuilhermBrSp I think we could do the rollback in this method. In my view, it's more contextualized. If you put it on onError, you may have to check the type of error to do rollback or not. I think it would be weird.